README.md
___This module has been superseded by [`@google-cloud/common`](https://www.npmjs.com/package/@google-cloud/common).___

Source code for `@google-cloud/common` is available in the [google-cloud-node](https://github.com/GoogleCloudPlatform/google-cloud-node/tree/master/packages/common) repository.


CHANGELOG.md
# Cloud Diagnostics Common ChangeLog

## 2017-09-05, Version 0.3.2 (Deprecated), @kjin

This module has been superseded by [@google-cloud/common](https://www.npmjs.com/package/@google-cloud/common).

## 2016-04-21, Version 0.3.1 (Experimental), @ofrobots

## 2016-10-24, Version 0.3.0 (Experimental), @matthewloring

### Commits

* [[`527d1fc183`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/527d1fc183)] - Implemented reading key file/credentials config object for auth client (#28) (Kelvin Jin) [#28](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/pull/28)
* [[`fbe5f860fa`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/fbe5f860fa)] - Update to released auth library (#29) (Matthew Loring) [#29](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/pull/29)

## 2016-10-03, Version 0.2.5 (Experimental), @matthewloring

### Commits

* [[`c6c67a0bef`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/c6c67a0bef)] - **deps**: pick-up symlink handling fix in auth (#25) (Ali Ijaz Sheikh) [#25](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/pull/25)
* [[`eb64896ad5`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/eb64896ad5)] - Add badges and update dependencies (#24) (Matthew Loring) [#24](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/pull/24)
* [[`fcbcee19e6`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/fcbcee19e6)] - Update README.md (#23) (Cristian Cavalli)
* [[`6b32250ddd`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/6b32250ddd)] - Increase test coverage >95% (#22) (Cristian Cavalli)

## 2016-04-18, Version 0.2.3 (Experimental), @matthewloring

### Commits

* [[`1c4a9ff72f`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/1c4a9ff72f)] - Get travis running (Matt Loring)
* [[`93811b80aa`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/93811b80aa)] - GCLOUD_PROJECT instead of GCLOUD_PROJECT_NUM (Matt Loring)
* [[`ad832dd61a`](https://github.com/GoogleCloudPlatform/cloud-diagnostics-common-nodejs/commit/ad832dd61a)] - Add silly log level above debug (Matt Loring)


test/test-utils.js
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
var proxyquire = require('proxyquire').noPreserveCache();
var nock = require('nock');
var assert = require('assert');
var path = require('path');

function GoogleAuth() {}
GoogleAuth.prototype.getApplicationDefault = function(cb) {
  return cb(null, 'Awesome Auth Client');
};
GoogleAuth.prototype.fromJSON = function(json, cb) {
  return cb(null, json);
};
GoogleAuth.prototype.fromStream = function(stream, cb) {
  var contents = '';
  stream.on('data', function(data) {
    contents += data;
  });
  stream.on('end', function() {
    return cb(null, JSON.parse(contents));
  });
};

var utils = proxyquire('../lib/utils.js', {
  'google-auth-library': GoogleAuth
});

nock.disableNetConnect();

describe('utils', function() {
  after(function() {
    nock.enableNetConnect();
  });

  describe('getProjectNumber', function() {

    it('should be able to get project number from metadata service',
      function(done) {
        var scope = nock('http://metadata.google.internal')
                      .get('/computeMetadata/v1/project/numeric-project-id')
                      .reply(200, '567');
        utils.getProjectNumber(function(err, project) {
          assert.ok(!err);
          assert.strictEqual(project, '567');
          scope.done();
          done();
        });
      });
    
    it('should be able to handle 500\'s from the metadata service',
      function(done) {
        var scope = nock('http://metadata.google.internal')
                      .get('/computeMetadata/v1/project/numeric-project-id')
                      .reply(500, {error: true});
        utils.getProjectNumber(function(err, project) {
          assert.strictEqual(typeof err, 'object');
          assert.ok(err instanceof Error);
          assert.strictEqual(err.message, 'Error discovering project num');
          assert.strictEqual(project, undefined);
          scope.done();
          done();
        });
      });

    it('should accept an optional headers parameter', function(done) {
      var scope =
        nock('http://metadata.google.internal', {
            reqheaders: {'Flux': 'Capacitor'}
          })
          .get('/computeMetadata/v1/project/numeric-project-id')
          .reply(200, '789');
      utils.getProjectNumber({'Flux': 'Capacitor'}, function(err, project) {
        assert.ok(!err);
        assert.strictEqual(project, '789');
        scope.done();
        done();
      });
    });

    it('Should callback with ENOTFOUND', function (done) {
      var oldEnv = process.env.GCLOUD_PROJECT;
      process.env.GCLOUD_PROJECT = './this-should-not-exist.json';
      var scope = nock('http://metadata.google.internal')
        .get('/computeMetadata/v1/project/numeric-project-id')
        .once()
        .replyWithError({'message': 'Not Found', code: 'ENOTFOUND'});
      utils.getProjectNumber(function (e, result) {
        assert.ok(e instanceof Error, 'e should be an instance of Error');
        assert.deepEqual(result, null);
        process.env.GCLOUD_PROJECT = oldEnv;
        scope.done();
        done();
      });
    });
  });

  describe('getProjectId', function() {

    it('should be able to get project id from metadata service',
      function(done) {
        var scope = nock('http://metadata.google.internal')
                      .get('/computeMetadata/v1/project/project-id')
                      .reply(200, 'a-stub-project-id');
        utils.getProjectId(function(err, projectId) {
          assert.ok(!err);
          assert.strictEqual(projectId, 'a-stub-project-id');
          scope.done();
          done();
        });
      });
    
    it('should be able handle 500\'s from the service',
      function(done) {
        var scope = nock('http://metadata.google.internal')
                      .get('/computeMetadata/v1/project/project-id')
                      .reply(500, {error: true});
        utils.getProjectId(function(err, projectId) {
          assert.strictEqual(typeof err, 'object');
          assert.ok(err instanceof Error);
          assert.strictEqual(err.message, 'Error discovering project id');
          assert.strictEqual(projectId, null);
          scope.done();
          done();
        });
      });

    it('should accept an optional headers parameter', function(done) {
      var scope =
        nock('http://metadata.google.internal', {
            reqheaders: {'Flux': 'Capacitor'}
          })
          .get('/computeMetadata/v1/project/project-id')
          .reply(200, 'a-stub-project-id');
      utils.getProjectId({'Flux': 'Capacitor'}, function(err, project) {
        assert.ok(!err);
        assert.strictEqual(project, 'a-stub-project-id');
        scope.done();
        done();
      });
    });

    it('Should callback with ENOTFOUND', function (done) {
      var oldEnv = process.env.GCLOUD_PROJECT;
      process.env.GCLOUD_PROJECT = './this-should-not-exist.json';
      var scope = nock('http://metadata.google.internal')
        .get('/computeMetadata/v1/project/project-id')
        .once()
        .replyWithError({'message': 'Not Found', code: 'ENOTFOUND'});
      utils.getProjectId(function (e, result) {
        assert.ok(e instanceof Error, 'e should be an instance of Error');
        assert.deepEqual(result, null);
        process.env.GCLOUD_PROJECT = oldEnv;
        scope.done();
        done();
      });
    });
  });

  describe('authorizedRequestFactory', function() {
    it('should return a function', function() {
      var result = utils.authorizedRequestFactory(['fake-scope']);
      assert(typeof result === 'function');
    });
  });

  describe('getAuthClient', function() {
    // require is relative to this file
    var credentials = require('./fixtures/stub_cert.json');
    var keyFilePath = path.join('test', 'fixtures', 'stub_cert.json');

    it('should work with empty scopes', function(done) {
      utils.getAuthClient([], function(err, client) {
        assert(!err);
        assert(client);
        done();
      });
    });
    it('should call fromJSON when JSON credentials are provided',
      function(done) {
        var config = {
          credentials: credentials
        };
        utils.getAuthClient([], config, function(err, client) {
          assert(!err);
          assert(client);
          assert.equal(client, credentials);
          done();
        });
      });
    it('should call fromStream when JSON credentials path is provided',
        function(done) {
        var config = {
          keyFile: keyFilePath
        };
        utils.getAuthClient([], config, function(err, client) {
          assert(!err);
          assert(client);
          assert.deepEqual(client, credentials);
          done();
        });
      });
  });

  describe('requestWithRetry', function() {

    it('should not retry on successful request', function(done) {
      var attempt = 0;
      // a request that always succeeds
      var request = function(options, callback) {
        attempt += 1;
        callback(null, 'response', 'body');
      };
      utils.requestWithRetry(request, {},
        function(err, response, body) {
          assert.strictEqual(attempt, 1);
          done();
        });
      });

    it('should not retry on non-transient errors', function(done){
      var attempt = 0;
      // a request that always fails with HTTP 404
      var request = function(options, callback) {
        attempt += 1;
        callback({code: 404});
      };
      utils.requestWithRetry(request, {}, function(err, response, body) {
        assert.strictEqual(attempt, 1);
        assert.equal(err.code, 404);
        done();
      });
    });

    it('should retry atleast 4 times on unsuccessful requests', function(done){
      this.timeout(30000); // this test takes a long time
      var attempt = 0;
      // a request that always fails
      var request = function(options, callback) {
        attempt += 1;
        callback({code: 429});
      };
      utils.requestWithRetry(request, {}, function(err, response, body) {
        assert.ok(attempt >= 4);
        done();
      });
    });

    it('should retry on errors 429, 500, 503', function(done) {
      this.timeout(30000); // this test takes a long time
      var attempt = 0;
      var request = function(options, callback) {
        attempt += 1;
        switch (attempt) {
          case 1:
            callback({code: 429});
            return;
          case 2:
            callback({code: 500});
            return;
          case 3:
            callback({code: 503});
            return;
          case 4:
            callback(null, 'response', 'body');
            return;
          default:
            assert.fail();
        }
      };

      utils.requestWithRetry(request, {}, function(err, response, body) {
        assert.ok(!err);
        assert.strictEqual(attempt, 4); // should have passed on 4th attempt
        done();
      });
    });
  });

  describe('getInstanceId - valid cases', function() {
    var STUB_ID = 'a-stub-instance-id';
    it(
      'Should be able to get the instance id without additional headers supplied',
      function (done) {
        var mock = nock('http://metadata.google.internal/computeMetadata/v1')
          .get('/instance/id')
          .once()
          .reply(200, STUB_ID);
        utils.getInstanceId(function (err, id) {
          assert.deepEqual(err, null, 'Error should be null');
          assert.deepEqual(STUB_ID, id, 'The id should be the stub id');
          mock.done();
          done();
        });
      }
    );
    it(
      'Should be able to get the instance id with additional headers supplied',
      function (done) {
        var mock = nock('http://metadata.google.internal/computeMetadata/v1',
          {reqHeaders: {'x-custom-header': 'true'}})
          .get('/instance/id')
          .once()
          .reply(200, STUB_ID);
        utils.getInstanceId({'x-custom-header': 'true'}, function (err, id) {
          assert.deepEqual(err, null, 'Error should be null');
          assert.deepEqual(STUB_ID, id, 'The id should be the stub id');
          mock.done();
          done();
        });
      }
    );
  });

  describe('getHostname - valid cases', function() {
    var STUB_HOSTNAME = 'a-stub-hostname';
    it(
      'Should be able to get the hostname without additional headers supplied',
      function (done) {
        var mock = nock('http://metadata.google.internal/computeMetadata/v1')
          .get('/instance/hostname')
          .once()
          .reply(200, STUB_HOSTNAME);
        utils.getHostname(function (err, id) {
          assert.deepEqual(err, null, 'Error should be null');
          assert.deepEqual(STUB_HOSTNAME, id,
            'The hostname should be the stub hostname');
          mock.done();
          done();
        });
      }
    );
    it(
      'Should be able to get the hostname with additional headers supplied',
      function (done) {
        var mock = nock('http://metadata.google.internal/computeMetadata/v1',
          {reqHeaders: {'x-custom-header': 'true'}})
          .get('/instance/hostname')
          .once()
          .reply(200, STUB_HOSTNAME);
        utils.getHostname({'x-custom-header': 'true'}, function (err, id) {
          assert.deepEqual(err, null, 'Error should be null');
          assert.deepEqual(STUB_HOSTNAME, id,
            'The hostname should be the stub hostname');
          mock.done();
          done();
        });
      }
    );
  });
});


test/test-logger.js
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';
var os = require('os');
var fs = require('fs');
var fsMock = require('mock-fs');
var proxyquire = require('proxyquire').noPreserveCache();
var assert = require('assert');
var PREFIX = 'mock_fd_logger';
var MOCK_FILE = [PREFIX, 'log.txt'].join('_');
var MOCK_PATH = os.tmpdir()+'/'+MOCK_FILE; // mock-fs always uses forward slashes
var SECONDS = 1471332553; // Some date value measured in seconds from epoch.

/* Stub console.log for testing */
var buffer = [];
var orig = console._stdout.write;
console._stdout.write = function() {
  buffer.push(arguments[0]);
  orig.apply(this, arguments);
};

describe('mocking fd logger with config', function () {
  it('should write an entry to the fs mock when log is called', function () {
    var logger = require('../lib/logger.js');
    var l = logger.create(logger.SILLY, PREFIX, 9);
    for (var i = 0; i < 15; i += 1) {
      l.info('info_');
    }
    var s = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual(s.split('\n').length, 10, 'There should be 10 entries in the logs')
  });
  before(function () {
    var mockTree = {};
    mockTree[os.tmpdir()] = {};
    mockTree[os.tmpdir()][MOCK_FILE] = '';
    fsMock(mockTree, {createTmp: false});
  });
  after(function () {
    fsMock.restore();
  });
});

describe('mocking fd logger without config', function () {
  it('should not write an entry to the fs mock when log is called', function () {
    var logger = require('../lib/logger.js');
    var l = logger.create();
    for (var i = 0; i < 5; i += 1) {
      l.info('info_');
    }
    var s = fs.readFileSync(os.tmpdir()+'/_log.txt').toString();
    assert.deepEqual(s, '', 'The log file should be empty');
  });
  before(function () {
    var mockTree = {};
    mockTree[os.tmpdir()] = {};
    mockTree[os.tmpdir()]['_log.txt'] = '';
    fsMock(mockTree, {createTmp: false});
  });
  after(function () {
    fsMock.restore();
  });
});

describe('logging a breakpoint object', function () {
  var checkTime = /createdTime/;
  var checkCondition = /condition/;
  var checkExpression = /expressions/;
  it('should log a breakpoint object correctly', function () {
    var logger = require('../lib/logger.js');
    var correctOutput = [
      'ERROR:mock_fd_logger: breakpoint id: 0,\n\tlocation: { line: 3, path:',
      ' \'/my/project/root/test/fixtures/a/hello.js\' }\n\tcreatedTime: ',
      new Date(SECONDS * 1000).toString(),
      '\n\tcondition: \'if n == 3 then true else false\'\n\texpressions: ',
      '[ \'if n == 3 then Math.PI * n else n\' ]\n'
    ].join('');
    var bp = {
      id: 0,
      location: {
        line: 3,
        path: '/my/project/root/test/fixtures/a/hello.js'
      },
      createdTime: {seconds: SECONDS},
      condition: 'if n == 3 then true else false',
      expressions: ['if n == 3 then Math.PI * n else n']
    };
    var l = logger.create(logger.SILLY, PREFIX, 10);
    l.breakpoint(logger.ERROR, '', bp);
    var s = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual(s, correctOutput, 'The log break point should be properly formatted');
  });
  it('should log a breakpoint without time correctly', function () {
    var log, writer;
    var logger = require('../lib/logger.js');
    var bp = {
      id: 0,
      location: {
        line: 3,
        path: '/my/project/root/test/fixtures/a/hello.js'
      },
      condition: 'if n == 3 then true else false',
      expressions: ['if n == 3 then Math.PI * n else n']
    };
    writer = logger.create(logger.SILLY, PREFIX, 10);
    writer.breakpoint(logger.ERROR, '', bp);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.ok(!checkTime.test(log));
    assert.ok(checkCondition.test(log));
    assert.ok(checkExpression.test(log));
  });
  it('should log a breakpoint without condition correctly', function () {
    var log, writer;
    var logger = require('../lib/logger.js');
    var bp = {
      id: 0,
      location: {
        line: 3,
        path: '/my/project/root/test/fixtures/a/hello.js'
      },
      createdTime: {seconds: SECONDS},
      expressions: ['if n == 3 then Math.PI * n else n']
    };
    writer = logger.create(logger.SILLY, PREFIX, 10);
    writer.breakpoint(logger.ERROR, '', bp);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.ok(checkTime.test(log));
    assert.ok(!checkCondition.test(log));
    assert.ok(checkExpression.test(log));
  });
  it('should log a breakpoint without expressions correctly', function () {
    var log, writer;
    var logger = require('../lib/logger.js');
    var bp = {
      id: 0,
      location: {
        line: 3,
        path: '/my/project/root/test/fixtures/a/hello.js'
      },
      createdTime: {seconds: SECONDS},
      condition: 'if n == 3 then true else false'
    };
    writer = logger.create(logger.SILLY, PREFIX, 10);
    writer.breakpoint(logger.ERROR, '', bp);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.ok(checkTime.test(log));
    assert.ok(checkCondition.test(log));
    assert.ok(!checkExpression.test(log));
  });
  it('should not log a breakpoint if the logging level is set lower', function () {
    var log, writer;
    var logger = require('../lib/logger.js');
    var bp = {
      id: 0,
      location: {
        line: 3,
        path: '/my/project/root/test/fixtures/a/hello.js'
      },
      createdTime: {seconds: SECONDS},
      condition: 'if n == 3 then true else false'
    };
    writer = logger.create(logger.ERROR, PREFIX, 10);
    writer.breakpoint(logger.DEBUG, '', bp);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual('', log, 'The log should be empty');
  });
  it('should log an array of break points', function () {
    var correctOutput = [
      'ERROR:mock_fd_logger: \nERROR:mock_fd_logger: breakpoint id: 0,',
      '\n\tlocation: { line: 3, path: \'/my/project/root/test/fixtures/',
      'a/hello.js\' }\n\tcreatedTime: ',
      new Date(SECONDS*1000).toString(),
      '\n\tcondition: \'',
      'if n == 3 then true else false\'\nERROR:mock_fd_logger: breakpoint ',
      'id: 1,\n\tlocation: { line: 10,\n  path: ',
      '\'/my/project/root/test/fixtures/a/goodbye.js\' }\n\tcreatedTime: ',
      new Date(SECONDS*1000).toString(),
      '\n\tcondition: \'if n == 4 then true else false\'\n'
    ].join('');
    var log, writer;
    var logger = require('../lib/logger.js');
    var bps = [
      {
        id: 0,
        location: {
          line: 3,
          path: '/my/project/root/test/fixtures/a/hello.js'
        },
        createdTime: {seconds: SECONDS},
        condition: 'if n == 3 then true else false'
      },
      {
        id: 1,
        location: {
          line: 10,
          path: '/my/project/root/test/fixtures/a/goodbye.js'
        },
        createdTime: {seconds: SECONDS},
        condition: 'if n == 4 then true else false'
      }
    ];
    writer = logger.create(logger.DEBUG, PREFIX, 10);
    writer.breakpoints(logger.SILLY, '', bps);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual('', log, 'The log should be empty');
    writer.breakpoints(logger.ERROR, '', bps);
    // cristiancavalli added the \r\n replace with \n since it looks
    // like utils.inspect is changing behavior in node v6.4.0 and
    // newlining with \r\n instead of just \n
    // @TODO figure out if this is intended and why - the rest of the
    // ouput uses /n and other versions pass without this
    log = fs.readFileSync(MOCK_PATH).toString().replace(/\r\n/g, '\n');
    assert.deepEqual(correctOutput, log, 'The log should be properly formatted');
  });
  it('should log an interval', function () {
    var log, writer;
    var correctOutput = 'ERROR:mock_fd_logger: test 100000.0002ms\n';
    var logger = require('../lib/logger.js');
    writer = logger.create(logger.ERROR, PREFIX, 10);
    writer.interval(logger.DEBUG, 'test', 1000);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual('', log, 'The log should be empty');
    writer.interval(logger.ERROR, 'test', [100, 200]);
    log = fs.readFileSync(MOCK_PATH).toString();
    assert.deepEqual(correctOutput, log, 'The log should be properly formatted');
  });
  beforeEach(function () {
    var mockTree = {};
    mockTree[os.tmpdir()] = {};
    mockTree[os.tmpdir()][MOCK_FILE] = '';
    fsMock(mockTree, {createTmp: false});
  });
  afterEach(function () {
    fsMock.restore();
  });
});

describe('logger base-functionality', function() {
  var logger = require('../lib/logger.js');
  it('should return a logger through the create function', function() {
    var l = logger.create(logger.DEBUG, __filename);
    assert.ok(l);
    assert.ok(l.error);
    assert.ok(typeof l.error === 'function');
  });

  it('should generate log messages successfully', function() {
    var l = logger.create(logger.SILLY, 'foobar');
    l.error('a');
    assert.ok(/ERROR.*: a/.test(buffer.pop()));
    l.warn('b', 'c');
    assert.ok(/WARN.*: b c/.test(buffer.pop()));
    l.info('d', 'e', 'f');
    assert.ok(/INFO.*: d e f/.test(buffer.pop()));
    l.debug('g');
    assert.ok(/DEBUG.*: g/.test(buffer.pop()));
    l.silly('h');
    assert.ok(/SILLY.*: h/.test(buffer.pop()));
  });

  it('should not log when the default level is lower', function() {
    buffer = [];

    var l = logger.create(logger.WARN, 'foobar');

    l.error('a');
    assert.ok(buffer.length === 1);
    buffer.pop();

    l.warn('b');
    assert.ok(buffer.length === 1);
    buffer.pop();

    l.info('c');
    assert.ok(buffer.length === 0);

    l.debug('d');
    assert.ok(buffer.length === 0);
  });
});


lib/utils.js
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 'use strict';

var fs = require('fs');
var GoogleAuth = require('google-auth-library');
var request = require('request');
// we only need a single instance
var googleAuth = new GoogleAuth();

/** @const {number} */ var MAX_RETRY_ATTEMPTS = 5;
/** @const {number} */ var MIN_RETRY_TIMEOUT = 1000; // milliseconds

var METADATA_URL = 'http://metadata.google.internal/computeMetadata/v1';

/**
 * Returns true if `err` is a transient error code.
 * @param {?number} err The error code.
 * @return {boolean} Whether `err` is a transient error.
 */
function isTransientError(err) {
  // 429 - Too many requests.
  // 500 - Internal server error.
  // 503 - Service Unavailable.
  return (err && [429, 500, 503].indexOf(err.code) !== -1);
}

/**
 * Returns a backoff delay using an exponential backoff algorithm.
 * @param {number} attempt 1-indexed attempt number. The first retry would
 *    be attempt number 2.
 * @return {number} backoff delay in milliseconds.
 */
function retryDelay(attempt) {
  return MIN_RETRY_TIMEOUT * Math.pow(2, (attempt-1));
}

/**
 * Returns a google auth client for the current application with provided
 * scopes and configuration .
 */
function getAuthClient(scopes, config, callback) {
  if (typeof(config) === 'function') {
    callback = config;
    config = null;
  }

  if (config && config.keyFile) {
    googleAuth.fromStream(fs.createReadStream(config.keyFile), addScope);
  } else if (config && config.credentials) {
    googleAuth.fromJSON(config.credentials, addScope);
  } else {
    googleAuth.getApplicationDefault(addScope);
  }

  function addScope(err, authClient) {
    if (err) {
      callback(err);
      return;
    }
    if (authClient.createScopedRequired && authClient.createScopedRequired()) {
      authClient = authClient.createScoped(scopes);
    }
    callback(null, authClient);
  }
}

/**
 * Performs the provided request fn using the options and callback. If the
 * request fails with a server error, it automatically retries using exponential
 * backoff. This will retry atleast 4 times.
 *
 * TODO(ofrobots): maybe accept a config object instead of the request function?
 *     Perhaps we can allow the caller to specify retry count, etc.
 *
 * @param {function(Object, function(=?,=?,=?):?} request style function
 *     accepting (options, callback).
 * @param {Object} options options to pass to request function
 * @param {Function} callback for request
 */
function requestWithRetry(request, options, callback) {
  function tryRequest(attempt) {
    request(options, function(err, response, body) {
      if (isTransientError(err) && attempt < MAX_RETRY_ATTEMPTS) {
        var delay = retryDelay(attempt);
        setTimeout(function() {
          tryRequest(attempt + 1);
        }, delay);
        return;
      }
      // not a (server) error, or retried too many times already.
      callback(err, response, body);
    });
  }

  tryRequest(1);
}

/**
 * Returns a request style function that can make authorized requests to a
 * Google API using Google Application Default credentials. This hides the
 * the details of working with auth in the client code.
 *
 * @param {Array<string>} scopes list of scopes to request as part of auth
 * @param {Object} config an object with extra configuration parameters (such
 *     as keyFile or key)
 * @return {function(Object, function(=?,=?,=?):?)} request style function
 *     accepting (options, callback)
 */
function authorizedRequestFactory(scopes, config) {
  // The AuthClient instance associated with each instantiation
  var authClient;
  function makeRequest(options, callback) {
    // authClient expects options to be an object rather than a bare url.
    // Coerce into an object here
    if (typeof options === 'string') {
      options = {url: options};
    }
    if (authClient) {
      authClient.request(options, function(err, body, response) {
        // Ugh. google-auth-library changes the argument order for the
        // callback. Fix that here.
        callback(err, response, body);
      });
    } else {
      getAuthClient(scopes, config, function(err, client) {
        if (err) {

          callback(err);
          return;
        }
        if (!authClient) {
          // google-auth-library changes the argument order for the
          // callback. Fix that here.
          authClient = client;
        }
        authClient.request(options, function(err, body, response) {
          callback(err, response, body);
        });
      });
    }
  }

  return function(options, callback) {
    requestWithRetry(makeRequest, options, callback);
  };
}

function getMetadataValue(url, headers, callback) {
  headers['Metadata-Flavor'] = 'Google';

  requestWithRetry(request, {
    url: url,
    headers: headers,
    method: 'GET'
  }, callback);
}

/**
 * Attempts to retrieve the project number for the current active project from
 * the metadata service (See https://cloud.google.com/compute/docs/metadata).
 *
 * @param {object=} headers optional headers to include in the http request.
 *     Note that the headers, if provided, may be extended with extra
 *     properties.
 * @param {function(?, number):?} callback an (err, result) style callback
 */
function getProjectNumber(headers, callback) {
  if (typeof headers === 'function') {
    callback = headers;
    headers = {};
  }
  getMetadataValue(METADATA_URL + '/project/numeric-project-id',
      headers, function(err, response, project) {
    if (!err && response.statusCode === 200) {
      return callback(null, project);
    } else if (err && err.code === 'ENOTFOUND') {
      return callback(new Error('Could not auto-discover project-id. Please export ' +
        'GCLOUD_PROJECT with your project name'));
    } else {
      return callback(err || new Error('Error discovering project num'));
    }
  });
}

/**
 * Attempts to retreive the project id for the current active project from the
 * metadata service. The GCLOUD_PROJECT env variable or another identifying
 * project name/id must be set so that the underlying request library can
 * successfully query the metadata service.
 * {@link https://cloud.google.com/compute/docs/storing-retrieving-metadata}
 * @param {Object} [headers] - An optional set of headers to include in the http
 *  request. This function may mutate the given headers object.
 * @param {getProjectIdCallback} callback - A callback to receive the
 *  response body (project id) or error encountered during the request.
 */
function getProjectId(headers, callback) {
  if (typeof headers === 'function') {
    callback = headers;
    headers = {};
  }
  getMetadataValue(METADATA_URL + '/project/project-id', headers,
    function (err, response, projectId) {
      if (!err && response.statusCode === 200) {
        return callback(null, projectId);
      } else if (err && err.code === 'ENOTFOUND') {
        return callback(new Error('Could not auto-discover project-id.' +
          'Please export GCLOUD_PROJECT with your project name'), null);
      }
      return callback(err || new Error('Error discovering project id'), null);
  });
}

/**
 * Callback for getProjectId function. This callback will be invoked once
 * negotiation with the project-id metadata endpoint either succeeds or fails.
 * Always calls back with two parameters, one of which will always be null.
 * @callback getProjectIdCallback
 * @param {Error|Null} - If an error is encountered during the request this
 *  param will be an instance of the Error class otherwise it will be null.
 * @param {String|Null} - If no error is encountered during the request then
 *  this param will be of type string and its value will be the related project
 *  id; otherwise it will be null.
 * @example
 * function myCallback (err, projectId) {
 *  if (err) {
 *    console.error('Encountered error fetching project id', err);
 *    return;
 *  }
 *  console.log('Got project id', projectId);
 * }
 * utils.getProjectId(myCallback);
 */

/**
 * Attempts to retrieve the GCE instance hostname for the current active project
 * from the metadata service (See https://cloud.google.com/compute/docs/metadata).
 *
 * @param {object=} headers optional headers to include in the http request.
 *     Note that the headers, if provided, may be extended with extra
 *     properties.
 * @param {function(?, number):?} callback an (err, result) style callback
 */
function getHostname(headers, callback) {
  if (typeof headers === 'function') {
    callback = headers;
    headers = {};
  }
  getMetadataValue(METADATA_URL + '/instance/hostname',
      headers, function(err, response, hostname) {
    callback(err, hostname);
  });
}

/**
 * Attempts to retrieve the GCE instance id for the current active project
 * from the metadata service (See https://cloud.google.com/compute/docs/metadata).
 *
 * @param {object=} headers optional headers to include in the http request.
 *     Note that the headers, if provided, may be extended with extra
 *     properties.
 * @param {function(?, number):?} callback an (err, result) style callback
 */
function getInstanceId(headers, callback) {
  if (typeof headers === 'function') {
    callback = headers;
    headers = {};
  }
  getMetadataValue(METADATA_URL + '/instance/id',
      headers, function(err, response, id) {
    callback(err, id);
  });
}

module.exports = {
  getProjectNumber: getProjectNumber,
  getProjectId: getProjectId,
  getHostname: getHostname,
  getInstanceId: getInstanceId,
  authorizedRequestFactory: authorizedRequestFactory,
  requestWithRetry: requestWithRetry,
  getAuthClient: getAuthClient
};
